From patchwork Wed May 12 01:47:57 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12252649
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 7A5BFC433B4
	for <kvm@archiver.kernel.org>; Wed, 12 May 2021 03:58:58 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 566CD61205
	for <kvm@archiver.kernel.org>; Wed, 12 May 2021 03:58:58 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S229597AbhELEAE (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Wed, 12 May 2021 00:00:04 -0400
Received: from aserp2130.oracle.com ([141.146.126.79]:51640 "EHLO
        aserp2130.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S229447AbhELEAD (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 12 May 2021 00:00:03 -0400
Received: from pps.filterd (aserp2130.oracle.com [127.0.0.1])
        by aserp2130.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 14C2ZY7E029450;
        Wed, 12 May 2021 02:37:13 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=T6DXo5D3HpC7gxwXj2dOrWQonakJF0RILsSsA4/u6Tk=;
 b=TQQ4PRrom+kBoeEN/eh4/a87tlsCEt0gEYOA4IslF4OTv6Q6vCatPL5gQRPjaOOol7Pu
 eWlyv/GE4m2wUB97RXHV5L9On3IUpWGMHF3fVCCFIrNU+J1+/e14jM3t3+AwZYxgd9hN
 DzXZT8FXI+CE3FddN0Nt2Jq7eWTcFmpZLBLyj71kCim+En5ozWidVsOXnW++g2WvFx4L
 x7vHrr1ggWTlbnVDhFGaFmvj/0JAebKU/+nXSkJ+LNCgbhFsFCMeTbMgyWJAEcLvVD3M
 /pu89sEkepScCwtDG335gnFY3aQ3GjNf+/BmyZFHYsBqafzKx0XXqfniPyyn3PYJlV4m dA==
Received: from aserp3020.oracle.com (aserp3020.oracle.com [141.146.126.70])
        by aserp2130.oracle.com with ESMTP id 38dg5bgrn5-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 12 May 2021 02:37:13 +0000
Received: from pps.filterd (aserp3020.oracle.com [127.0.0.1])
        by aserp3020.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 14C2YgI4168780;
        Wed, 12 May 2021 02:37:12 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2176.outbound.protection.outlook.com [104.47.55.176])
        by aserp3020.oracle.com with ESMTP id 38djfadpax-2
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 12 May 2021 02:37:12 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=gHCr0jyku5Mco1YF6HP7pziGaL0GtA+hoBgxnmZFQJSYwtcmxLdWp47SlYjVyPgYeswzDepXIHPtwOTSPmQIIBkEixOqu+y6Nl01szdU60KQ+8UETjxM+syWVkf0y4PLaw6f0RgT8SdwSZzP+nuYEScPCnLwLtf6NHgBoY4v/zw4L7W0f86gsNV3y99G1lGjsdnQ+wTRHzwopavOrlxSoieSIJVz2xhK622Rwo6mQ6sm1F09ykta2GVwsDw5leTf1DbRnF/ijcDwojFgOzO6RdwY1EFTrBr9UagSqSzsHFrFCQJv4jj9c8XQ9cTFP/QN1T81rbOhowdoxSGgl8zzRA==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=T6DXo5D3HpC7gxwXj2dOrWQonakJF0RILsSsA4/u6Tk=;
 b=OzaACOOGJhbUYo6GOLQBJumKlF9tyfkrDdeGedy8fyjrQN+bMHYFsFXG1QMIagqGQvoAgStM2Mo/jGsOaP4WisMoB5o0v4eAEhPBMVQTKJz45ZOUnSw37USxGZkD0eIbV+QF4FehtR8bYUB++Sd/DSgVovsJizrv2oEL1Yb9nIDapdbfj7dcFZvzGLUeSOqbBw0sXDWnYKRYmqbl6vdwPAW7T1Qr/OQtYJf77hU7gqMR8qt+KaYokXdmXUx6SpfiS3torN+nRLqPn27x8Yz6HsEIkQp5RWF5SCnO4ncOi5popLEYW/5GsyhH055GD1RhAJNynNpfMIuXRwrQ7a/8Kw==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=T6DXo5D3HpC7gxwXj2dOrWQonakJF0RILsSsA4/u6Tk=;
 b=PXpdh8wAskDbR1xBps/IFXvCWoONKWuApkEJudT2KZCsoV7PaIXuNmlj/0/pcdimYnOjB/dK3Uq0z88/dcqn5MbLgI2MMn21GX59okbPSOCiEsCZQ7HRT0DnOybjtm22EGT0xAutRSu9v1e6wOj+WGyMv+U28kEWoTiibN4p2y0=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4425.namprd10.prod.outlook.com (2603:10b6:806:11b::9) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4129.26; Wed, 12 May
 2021 02:37:11 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::5911:9489:e05c:2d44]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::5911:9489:e05c:2d44%5]) with mapi id 15.20.4108.031; Wed, 12 May 2021
 02:37:11 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 1/3] KVM: nVMX: Move 'nested_run' counter to
 enter_guest_mode()
Date: Tue, 11 May 2021 21:47:57 -0400
Message-Id: <20210512014759.55556-2-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210512014759.55556-1-krish.sadhukhan@oracle.com>
References: <20210512014759.55556-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.201.29]
X-ClientProxiedBy: BY5PR03CA0016.namprd03.prod.outlook.com
 (2603:10b6:a03:1e0::26) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.201.29) by
 BY5PR03CA0016.namprd03.prod.outlook.com (2603:10b6:a03:1e0::26) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4108.25 via Frontend
 Transport; Wed, 12 May 2021 02:37:10 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 5ab18957-70f2-441e-a896-08d914eed847
X-MS-TrafficTypeDiagnostic: SA2PR10MB4425:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB442555D7DCE9E9117210505D81529@SA2PR10MB4425.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:167;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 7WfyLZIoF8JTblOjz4u0cJ1U5dJOGR5mZ2OFwn06mn3EJolQCc/wv6cipgy9frq5NWHX1sEeYNpl3sl9fMNFeNtWKiixLq5JwKWZUeA5YaMhipT1bWQxmmnO5ODzlhNf1+Wky5H0QnhlY6wUkt75CBa2bDXBAgjJgM4NAMsyL2Fd/BMEpuV/XNSXnzpeiB/gPeQJvHvYbSm7bXkvi2LfsjPVNEwjiRI2D6EyE5DIUAonsCl1SBARU3+T8DLBs8o7eqimxtAohpHyoHvDCTFnmfP22ABy7Pym9+I5dJxK+ak3U1rd/N7QvZarXZugzAXkLXWZHE9rCJ2vyVKXzMhquon7cISScGYL1DgoMHizac4vBU7m85CmvI3K1fiixS6J6jMeDnSD2M3vhGFjLJjDn0uGVPvlVYsMoDCjSRAB8fW0AdqEU3VCj/j1GLCahtZ7uvItf4au1mMPxd0JPk6Lmreehg8N6OM+iOS8YGt4Fc/u2+/Ifg554TJu/GkU/37SeABLfTwbhniJKT+VQaheMqQVLtTluPW6vYAwZD51So4TlLjr7+KdkFUcPgejUBg8pmt74NaR+oyskmhGz/0Fh1XfFuyqWSiSrEV8H31jrRDFeS0/HxL5CQQWxCzjBaoBbOgYewNbpHDSuDeEo9fqHg==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(346002)(136003)(39860400002)(366004)(396003)(376002)(66946007)(8936002)(66556008)(478600001)(6486002)(2906002)(1076003)(36756003)(5660300002)(83380400001)(186003)(16526019)(66476007)(316002)(86362001)(6666004)(52116002)(38100700002)(8676002)(26005)(4326008)(38350700002)(6916009)(2616005)(44832011)(7696005)(956004);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 CxkE+bVebwM3Vu7Z7vzAszxgq+9fWHogGmshp05Hiw0nzUleDispzVd5d1A3KicdCDvdixLOoxHXeWl/0FM3IDMXmZaShLxFBCwhzc2ooPNzjVM4z0rbGrnIhEqgdyMDs7O28NLq8G5wxQFK0TJqYJBvIyFqjqSxbWqBbMS02OYdgIjZ70js5o7eXut4RYpvbHZ/S4ZIBYldXmwDb3Gk0dfLLrZHkAUqTA/LyMXFEVxcUdQ1b55JM41jiOSPnp1mFktkwv+f5ooVDfocgchaFyIKUWZD+2LdhTJ1QR+GSPsutPf/CfwIikoNXJtEFknrrrfYfHblqBotDuvHh/TOlbDdabPxcIQOkc++p4l51umm8wbNPC2jWJ2DjzaAm4o5B93HUHoOi4xNvcBRSuReBgEdbyMfF97yImeSm6HDxSjS8WNaC4zOR2X0Bvawy1GhzXKO+IAeW0bdIVMRzmBhgsioHVIdExkFy1CcjhdssysP5lbdBNli4cw9YZ9H+4VWeY3avVwbxgsYfRz81NjZjqhDqqiJOOO01lRvkzY+cR/GkPDcuwemnQfsjAwrIU4+KP6jKlYGOOGpSTyHdcXeQcPFjxf9FgT/Ws6YhDQ6rk1ZhaxZTzU5OmcMyrwf6X/B0hQ/JSQiW1Y4NfCHQfraynj4GPXa4elRdUMG25j/SzCdzm++LGNDboVjwTU7sNrqxccTDwxGUQDnQgpgZiSUFfXBLiSTB07+PE5jM7Fsq+Q5FSyvv0chC5TNEqdUptvhdYXLbFkEXnRP21eV3rYy1eOStPAP8NMSlPCoTioF95is1LNOCLTqW3oTdC/Ugyie+4mOYChuvt+1piX+QBAVrOpCLhbaSvlx0DeXoA0eq/vLpFFQdd/BVW2X16CaPfQogu+Su8HOphVGSd3KmxJkbkMLLR1HDSpmJqHlHl8bdTRg6q0KOCF9CKjtumF4dptb4c3uBWOflWp9tSbocMT1vJ5GfO8tNbtUvJrOoM6C5lQEndjy9uSqgyXgw7YTi/3SUWbYLAa+CG0p+Boy0DeoiFeeTKQUFjQKczJYbfR4Zej6d4Mj8jqWjqzSCUb690+kRfFL5yaY4zFo+m2cFolaE3ucH5sevZ9/N6hDbfylH94D/Wu/WVHwDoBOttZswmE31kHj1Pkn333sJ/BLZlrFQo43LDsqUOkbwjs32bf7jTzQaloIkJzPPznT0ULoTIslg04vX9QSC8aUEJF0aP/bJGWhexthT++aFLYqHw90dC9lxs8lFUdUtJXabIwQYXps9gdDl+wyn2VQwwfCZ2fGAG8BYHgd8OBJ1oRQBZ9oS88FkmHX6vO9EM7NkCAzOyxn
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 5ab18957-70f2-441e-a896-08d914eed847
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 12 May 2021 02:37:11.5511
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 ItByUtnHszcqDynPqTbYWlNpiqM/jCuhX0//sKrina7LhbRvK+Xgz1uN3d9AsbR3djQ23NTaAw361ntio4K4kr0tHfxY8rE4NEpEmjz8hUk=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4425
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9981
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 bulkscore=0
 spamscore=0 mlxlogscore=999
 adultscore=0 phishscore=0 mlxscore=0 suspectscore=0 malwarescore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2104190000
 definitions=main-2105120018
X-Proofpoint-GUID: AQ7doeTnY4gdNViQQ8m7Uhss2P90yM6B
X-Proofpoint-ORIG-GUID: AQ7doeTnY4gdNViQQ8m7Uhss2P90yM6B
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9981
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0
 lowpriorityscore=0 malwarescore=0
 bulkscore=0 spamscore=0 clxscore=1015 priorityscore=1501 adultscore=0
 mlxlogscore=999 mlxscore=0 suspectscore=0 impostorscore=0 phishscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2104190000
 definitions=main-2105120018
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Move 'nested_run' counter to enter_guest_mode() because,
    i) This counter is common to both Intel and AMD and can be incremented
       from a common place,
    ii) guest mode is a more finer-grained state than the beginning of
	nested_svm_vmrun() and nested_vmx_run().

Also, rename it to 'nested_runs'.

Signed-off-by: Krish Sadhukhan <Krish.Sadhukhan@oracle.com>
---
 arch/x86/include/asm/kvm_host.h | 2 +-
 arch/x86/kvm/kvm_cache_regs.h   | 1 +
 arch/x86/kvm/svm/nested.c       | 2 --
 arch/x86/kvm/vmx/nested.c       | 2 --
 arch/x86/kvm/x86.c              | 2 +-
 5 files changed, 3 insertions(+), 6 deletions(-)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index 55efbacfc244..cf8557b2b90f 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1170,7 +1170,7 @@ struct kvm_vcpu_stat {
 	u64 req_event;
 	u64 halt_poll_success_ns;
 	u64 halt_poll_fail_ns;
-	u64 nested_run;
+	u64 nested_runs;
 	u64 directed_yield_attempted;
 	u64 directed_yield_successful;
 };
diff --git a/arch/x86/kvm/kvm_cache_regs.h b/arch/x86/kvm/kvm_cache_regs.h
index 3db5c42c9ecd..cf52cbff18d3 100644
--- a/arch/x86/kvm/kvm_cache_regs.h
+++ b/arch/x86/kvm/kvm_cache_regs.h
@@ -162,6 +162,7 @@ static inline u64 kvm_read_edx_eax(struct kvm_vcpu *vcpu)
 static inline void enter_guest_mode(struct kvm_vcpu *vcpu)
 {
 	vcpu->arch.hflags |= HF_GUEST_MASK;
+	++vcpu->stat.nested_runs;
 }
 
 static inline void leave_guest_mode(struct kvm_vcpu *vcpu)
diff --git a/arch/x86/kvm/svm/nested.c b/arch/x86/kvm/svm/nested.c
index 5e8d8443154e..34fc74b0d58a 100644
--- a/arch/x86/kvm/svm/nested.c
+++ b/arch/x86/kvm/svm/nested.c
@@ -596,8 +596,6 @@ int nested_svm_vmrun(struct kvm_vcpu *vcpu)
 	struct kvm_host_map map;
 	u64 vmcb12_gpa;
 
-	++vcpu->stat.nested_run;
-
 	if (is_smm(vcpu)) {
 		kvm_queue_exception(vcpu, UD_VECTOR);
 		return 1;
diff --git a/arch/x86/kvm/vmx/nested.c b/arch/x86/kvm/vmx/nested.c
index 6058a65a6ede..94f70c0af4a4 100644
--- a/arch/x86/kvm/vmx/nested.c
+++ b/arch/x86/kvm/vmx/nested.c
@@ -3454,8 +3454,6 @@ static int nested_vmx_run(struct kvm_vcpu *vcpu, bool launch)
 	u32 interrupt_shadow = vmx_get_interrupt_shadow(vcpu);
 	enum nested_evmptrld_status evmptrld_status;
 
-	++vcpu->stat.nested_run;
-
 	if (!nested_vmx_check_permission(vcpu))
 		return 1;
 
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 5bd550eaf683..6d1f51f6c344 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -243,7 +243,7 @@ struct kvm_stats_debugfs_item debugfs_entries[] = {
 	VCPU_STAT("l1d_flush", l1d_flush),
 	VCPU_STAT("halt_poll_success_ns", halt_poll_success_ns),
 	VCPU_STAT("halt_poll_fail_ns", halt_poll_fail_ns),
-	VCPU_STAT("nested_run", nested_run),
+	VCPU_STAT("nested_runs", nested_runs),
 	VCPU_STAT("directed_yield_attempted", directed_yield_attempted),
 	VCPU_STAT("directed_yield_successful", directed_yield_successful),
 	VM_STAT("mmu_shadow_zapped", mmu_shadow_zapped),

From patchwork Wed May 12 01:47:58 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12252737
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 1CC03C433B4
	for <kvm@archiver.kernel.org>; Wed, 12 May 2021 06:04:53 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id D26AD6162B
	for <kvm@archiver.kernel.org>; Wed, 12 May 2021 06:04:52 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S229730AbhELGF6 (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Wed, 12 May 2021 02:05:58 -0400
Received: from aserp2130.oracle.com ([141.146.126.79]:38408 "EHLO
        aserp2130.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S229580AbhELGF5 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 12 May 2021 02:05:57 -0400
Received: from pps.filterd (aserp2130.oracle.com [127.0.0.1])
        by aserp2130.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 14C2Xvsx028854;
        Wed, 12 May 2021 02:37:15 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=yns43+Cdy4SfJMq2HB6L/MW6Nm6xGP/To3F46na3ONg=;
 b=RtS5zRRv5WFGfZUwXxDIefwgmzSL9OXd8r40zvAMmsQgHb1ErzuALJVwQyJMsagWIgQF
 Y/4B0/u+IzawP1RwwGqAIS51h+3LAXbi2WHGL+rnqDOnSY/TbheqC/t+CccOT882f36v
 TQctUZotZ/KP+mthzTyzZoAEx7/Xc29molqpQfrC80LKttBcd7M1XXIm9jqNbpTXHWCs
 ujXHzHNbMx6GcWtZWOJaOTjhc7Upu6PZ47D7ARUBnkuQqkePgZ8BwIXdT6DDhQzU/3q9
 Iys1piVskK5ns2lf3BPfkns8MhmOgPZoh+2KwedZeprMVabkfEtQscjcvNvJRzDZPHn3 zw==
Received: from userp3020.oracle.com (userp3020.oracle.com [156.151.31.79])
        by aserp2130.oracle.com with ESMTP id 38dg5bgrn8-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 12 May 2021 02:37:15 +0000
Received: from pps.filterd (userp3020.oracle.com [127.0.0.1])
        by userp3020.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 14C2a2jq028938;
        Wed, 12 May 2021 02:37:14 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2177.outbound.protection.outlook.com [104.47.55.177])
        by userp3020.oracle.com with ESMTP id 38fh3xnkqu-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 12 May 2021 02:37:14 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=mW1J/2M2xsD44JAcVgv8ezgrxF6c3N2EQPNGVHych4r4BkWamImSxOZQEw70lQZkSM4ere7gmEWJHmd6HFMnITPhtYWYnArfr3cIMel0DJoeZj/7eCloimv1eNU6dDFNswQlH8+iQ3dRfgM3EdKlo1VXr2SgRpx0TzaTwGR0P6+Y2ln0JKlxR3hJVdJcL7nCojdmOrXixF9hXQGnDyo6m+uYM4UewbitS+W+1EOoHPukToEZxrMZY+ByacYlhXfycmndBEr1jVZGCJgIVRjrlRxm4EcL7OLNpyvPUxU/Mw5wJO/uRzqy9gSuBVWGumSbD3Wfuo6ZrZ5v9IO5216pQg==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=yns43+Cdy4SfJMq2HB6L/MW6Nm6xGP/To3F46na3ONg=;
 b=TnnLoC8SoZSORirbKW/aUM6fsM1Dk/fn+3y0tY2nZPBzgtpCGXPGFRhEpa2ru4eOjCtE46Cq0A5FReaCtSXVDosaGwXmdk6PXUAmbMeqdj4VIti6K1JiKvZpUkVZa3/7D1ueh6s49sZ9vOpB55OC/TNnQMSEGkKvJJdowA0/gYBeseJ0sjhFGLAtvMDlWSrkE1XTMdJyVzdjPR9CSootnMslhdy6+fDfenxVFTEYXfXjs0+keXyelMLi7nfif1VbOJrwoQkj7DFfFG1Sqt3XQ3IRU5rKYa8yPixKKquZMdiEd+cTBmGO/BXhwLgN/5Gnm8KeGXk7XRACPl5gcALTiA==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=yns43+Cdy4SfJMq2HB6L/MW6Nm6xGP/To3F46na3ONg=;
 b=OAOc/5n1+DPAn4FmSmFbur+gdaYq5Q/1nPnLbmntQ1kxsVnXy3GhAJqLEd6ViurLXo7TV0PyvoLdRK2IjTXq4dE6JTtjVDHJwjeJVC/1jhFVAbLSfNSlkRJCmXLncFrCwi0kHSssKalwr6t+B+00GWX4WBFjoKmLwnO7q8+ov6k=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4425.namprd10.prod.outlook.com (2603:10b6:806:11b::9) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4129.26; Wed, 12 May
 2021 02:37:12 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::5911:9489:e05c:2d44]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::5911:9489:e05c:2d44%5]) with mapi id 15.20.4108.031; Wed, 12 May 2021
 02:37:12 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 2/3] KVM: nVMX: Add a new VCPU statistic to show if VCPU is
 running nested guest
Date: Tue, 11 May 2021 21:47:58 -0400
Message-Id: <20210512014759.55556-3-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210512014759.55556-1-krish.sadhukhan@oracle.com>
References: <20210512014759.55556-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.201.29]
X-ClientProxiedBy: BY5PR03CA0016.namprd03.prod.outlook.com
 (2603:10b6:a03:1e0::26) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.201.29) by
 BY5PR03CA0016.namprd03.prod.outlook.com (2603:10b6:a03:1e0::26) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4108.25 via Frontend
 Transport; Wed, 12 May 2021 02:37:11 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 250bc023-dedd-4881-c6b5-08d914eed902
X-MS-TrafficTypeDiagnostic: SA2PR10MB4425:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB442528F31C8D77C3611CD22C81529@SA2PR10MB4425.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:6430;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 hlUCCR/JAC7T+rCLrEqNY3loJ4/o/I3ZnfvDW+55njs9siLYnU7fmGkfwukkwgSTbVaNbbEsJC5tbecAr3Ok8Up9Nr2K/1CJImFiW99lv7QsUcUPEPyn7jGSnu+HwC4M/dySx83u6bEmogLygIJhg84AZHjae90LitjICwBBIjqm/YjgRRL4py1MTzJrZUT7u+2mOPySGdcYnKjo8NvQCtk3mSkbLsWTZ5I5s45Pbe+haoghhM+HIcX7GbuT+BY9Ra6VwB1xtA4ie7YPHOj9OcPOTAtOV1xSEdccbH4GkW39iTnUVw4piVukyf1ba0K1V4SJ1LvuxRYzPpSW6hdSs7m6uWqf/grl2Sovqf91zPwlcCWEYpgAfx+Zfew1DYC5rY4D0F2eqqGULSh5ebbaact5tHwUz0iKsoLtvSarn5dWmYaTfPR4Z9T/8xtEzGQbuvn8/m6t6XeHRNaeR2bAFb/FiGTbyr/oA/EKvJt7UNHzdrX0QWXuQGj30hp8xpmhxXcHeXvCV83mpPG2KUjrY8TxnVbHhq+GACGU0nFUoCVqx8za88oT/KJgv351KQfwvWQgubiiNYciHfGBfl9ZqiaPBLYUuYyEXtwkv4qSrWM++kfBbD8NvnX4pFKijaC+R2JS16haS6n7ciVKGz5BtQ==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(346002)(136003)(39860400002)(366004)(396003)(376002)(66946007)(8936002)(66556008)(478600001)(6486002)(2906002)(1076003)(36756003)(5660300002)(83380400001)(186003)(16526019)(66476007)(316002)(86362001)(6666004)(52116002)(38100700002)(8676002)(26005)(4326008)(38350700002)(6916009)(2616005)(44832011)(7696005)(956004);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 XEETZBqXFuiXpAZ+cfacH+NRdn641h7rbSAHY+wRDL0DOy6yghXFzkk1MjzbN3dTbwrUUsVY8vAlGe/Xrc79cOpxlvR2yol2mP+K23VH2JEqNaOcZiT8qEHRoBoXWKtMbeOKfmI6q+atH5EayRR8JqyY80haH4NayTSJCIiHzqjVfYycKIYhRPbCUhNsLRtHzP5YOpczxGfelFlECPsYeUMFzPjZgN0ECHkMP1ZpXpzVpYbW2N4imX/pv7gJFbGCtSHOlC25Zg1hqY9gtroCNIDu7l1CO5KrdbdacLuqA2Xdd2Hj1Xjjw0psvgeC3C8oRSTls2xSNjLaxX/8wRFJp9UjSGcxq3mbYfG/gJXRyiw7vfUyeNq0cKB2/d/viU6QGJ1o8LFnW0o8sux27ApqeXo9bnZlttv3dovxr55ZMpua36Dbz4lY1AjCvr2b/Bnmi3KDVXZsu9K8ndfml7tVMKu9AfHtSan1Da8sgC1IpF9IzfIJ2Jf9fFFnhf0Ri1c3UrDEDzOZs+NSY54z1i9aov14rVBq29d+l9v2gytRThZNhjQvWIOVM1GcoPuO+wvPhx8YcpxVifskVrsZs6Oh5G500uYrS5xwB7c3xSicV3gPWl2iBz14xknS7tSluhZl/UV5aw5eiWqxTUBbw4LiMzlUdWuCrMSQFUTjiyDZmXeYOyxUr3nvpnwQitxo9MLmHMW/bwALunkSlox1dgNWCAQUzI7YrExP2SzYBFCWZJr0jUEA+nz3WrucH/MKfDjhE3i14SHJ03a2Au6SyKv6RU4SN9gD3s5ukw79LvV1I+DNLemVKHkpFN0ejSmDZ/A6lxLud//q7A+nLGq8TiNTUJnUlx/nFWS/dXqLUjHqnLVe0wpkkl80qJjM65YlRvG/Tlh9hic9p+WyghU5LVL5mwNYTw7hBPjHiXVqoUJor02vKJbjZ9d20nlT8sYqAD2Y9DQiPJA5EruQqRW1ZRTcV82Tcb2cgvWCAub9xZJxi5lvQQftQApSjBWvgE4vOEJJvOncTeWYzBTijZdkmu55Nrc3oo5Q72RwMETI6+rLm27C1EcnKcfM46bNQjFg96XsL3ITt3zr8IzUcF80Zpywc2Hxfrr+QPGZTMCV+jUSjSz+dkmUE6vVBESnOf7vJQamkMzCNvYUgaR6DDjSh3sYGn2Y7JQoSeJsiJGupZeCfrHBP/eusDzKkD3FsnmjrYrfc4264djiPUEDUZ0pzYhT2N7HI7b778I2UxCXnjTKF7SJ2hRQNTWm3f9l+cLUoHkIFpXLWKiNR7dBbr7oWvLMRA2vzyHOetFAEO2JsCQ2jOLvVgLLYOfvpsbyLuM9+utr
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 250bc023-dedd-4881-c6b5-08d914eed902
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 12 May 2021 02:37:12.7104
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 BFz0k+lST9vF48oz+J3/M11jMauOk2uIj+uIdC8Hr3AyWqeHBpG905Hg207/pdjX83Ac+BK1ROyP5vtQjgP5ymaFH3FasA41lJfKLBgG6Z0=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4425
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9981
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 spamscore=0
 phishscore=0 suspectscore=0
 malwarescore=0 adultscore=0 bulkscore=0 mlxlogscore=999 mlxscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2104190000
 definitions=main-2105120018
X-Proofpoint-GUID: j-oglSDhHeDuAOVlbWAEqCSh9P8j9t1q
X-Proofpoint-ORIG-GUID: j-oglSDhHeDuAOVlbWAEqCSh9P8j9t1q
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9981
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0
 lowpriorityscore=0 malwarescore=0
 bulkscore=0 spamscore=0 clxscore=1015 priorityscore=1501 adultscore=0
 mlxlogscore=999 mlxscore=0 suspectscore=0 impostorscore=0 phishscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2104190000
 definitions=main-2105120018
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Add the following per-VCPU statistic to KVM debugfs to show if a given
VCPU is running a nested guest:

	nested_guest_running

Also add this as a per-VM statistic to KVM debugfs to show the total number
of VCPUs running a nested guest in a given VM.

Signed-off-by: Krish Sadhukhan <Krish.Sadhukhan@oracle.com>
---
 arch/x86/include/asm/kvm_host.h |  1 +
 arch/x86/kvm/debugfs.c          | 11 +++++++++++
 arch/x86/kvm/kvm_cache_regs.h   |  3 +++
 arch/x86/kvm/x86.c              |  1 +
 4 files changed, 16 insertions(+)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index cf8557b2b90f..884f6e5ba669 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1173,6 +1173,7 @@ struct kvm_vcpu_stat {
 	u64 nested_runs;
 	u64 directed_yield_attempted;
 	u64 directed_yield_successful;
+	u64 nested_guest_running;
 };
 
 struct x86_instruction_info;
diff --git a/arch/x86/kvm/debugfs.c b/arch/x86/kvm/debugfs.c
index 7e818d64bb4d..465d243afaac 100644
--- a/arch/x86/kvm/debugfs.c
+++ b/arch/x86/kvm/debugfs.c
@@ -17,6 +17,15 @@ static int vcpu_get_timer_advance_ns(void *data, u64 *val)
 
 DEFINE_SIMPLE_ATTRIBUTE(vcpu_timer_advance_ns_fops, vcpu_get_timer_advance_ns, NULL, "%llu\n");
 
+static int vcpu_get_guest_mode(void *data, u64 *val)
+{
+	struct kvm_vcpu *vcpu = (struct kvm_vcpu *) data;
+	*val = vcpu->stat.nested_guest_running;
+	return 0;
+}
+
+DEFINE_SIMPLE_ATTRIBUTE(vcpu_guest_mode_fops, vcpu_get_guest_mode, NULL, "%lld\n");
+
 static int vcpu_get_tsc_offset(void *data, u64 *val)
 {
 	struct kvm_vcpu *vcpu = (struct kvm_vcpu *) data;
@@ -45,6 +54,8 @@ DEFINE_SIMPLE_ATTRIBUTE(vcpu_tsc_scaling_frac_fops, vcpu_get_tsc_scaling_frac_bi
 
 void kvm_arch_create_vcpu_debugfs(struct kvm_vcpu *vcpu, struct dentry *debugfs_dentry)
 {
+	debugfs_create_file("nested_guest_running", 0444, debugfs_dentry, vcpu,
+			    &vcpu_guest_mode_fops);
 	debugfs_create_file("tsc-offset", 0444, debugfs_dentry, vcpu,
 			    &vcpu_tsc_offset_fops);
 
diff --git a/arch/x86/kvm/kvm_cache_regs.h b/arch/x86/kvm/kvm_cache_regs.h
index cf52cbff18d3..348ef98c6199 100644
--- a/arch/x86/kvm/kvm_cache_regs.h
+++ b/arch/x86/kvm/kvm_cache_regs.h
@@ -163,6 +163,7 @@ static inline void enter_guest_mode(struct kvm_vcpu *vcpu)
 {
 	vcpu->arch.hflags |= HF_GUEST_MASK;
 	++vcpu->stat.nested_runs;
+	vcpu->stat.nested_guest_running = 1;
 }
 
 static inline void leave_guest_mode(struct kvm_vcpu *vcpu)
@@ -173,6 +174,8 @@ static inline void leave_guest_mode(struct kvm_vcpu *vcpu)
 		vcpu->arch.load_eoi_exitmap_pending = false;
 		kvm_make_request(KVM_REQ_LOAD_EOI_EXITMAP, vcpu);
 	}
+
+	vcpu->stat.nested_guest_running = 0;
 }
 
 static inline bool is_guest_mode(struct kvm_vcpu *vcpu)
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 6d1f51f6c344..01805b68dc9b 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -246,6 +246,7 @@ struct kvm_stats_debugfs_item debugfs_entries[] = {
 	VCPU_STAT("nested_runs", nested_runs),
 	VCPU_STAT("directed_yield_attempted", directed_yield_attempted),
 	VCPU_STAT("directed_yield_successful", directed_yield_successful),
+	VCPU_STAT("nested_guest_running", nested_guest_running),
 	VM_STAT("mmu_shadow_zapped", mmu_shadow_zapped),
 	VM_STAT("mmu_pte_write", mmu_pte_write),
 	VM_STAT("mmu_pde_zapped", mmu_pde_zapped),

From patchwork Wed May 12 01:47:59 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12252593
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 080F2C433ED
	for <kvm@archiver.kernel.org>; Wed, 12 May 2021 02:54:59 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id D3C9161944
	for <kvm@archiver.kernel.org>; Wed, 12 May 2021 02:54:58 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S229964AbhELC4F (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 11 May 2021 22:56:05 -0400
Received: from aserp2130.oracle.com ([141.146.126.79]:59688 "EHLO
        aserp2130.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S229932AbhELC4E (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 11 May 2021 22:56:04 -0400
Received: from pps.filterd (aserp2130.oracle.com [127.0.0.1])
        by aserp2130.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 14C2Y0C4028881;
        Wed, 12 May 2021 02:37:16 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=lYaG9mCpHdaSAqGNzPGndXNZxWV9gxCPg2KmqGBrnMA=;
 b=C9DAEPgNbmHkhJ5YUPLWBygaOD1aFJ8AK+Bt2l5TcghxGFw6TfAZyrc7flY6jD8o2Vpi
 upgZ29HQjE6T8gCFmmadM7UZ0LSKsJgNnWywVk21iGSLdi9dtqpxC8HASXB3/m4+TrCV
 eF3inFmw4QDLJ+J+MIdFzH0pJx7XEYjOhe6gLsL2aMGYXHXk8WvXySnbk3vkQV3rPXjE
 3LdabTkeA5zUk3GDpfWrK+IIL1JJ0SM31GzGbXVidobbh6fiASi20XtYZD0RiTNHz0MB
 Hcrm5S8ywd6n+SVTwH+yh72tENICt0UmwXVYRLK1rZHUeJ7NEzpa1ZEKyMD13KaRGYs0 nA==
Received: from userp3020.oracle.com (userp3020.oracle.com [156.151.31.79])
        by aserp2130.oracle.com with ESMTP id 38dg5bgrn9-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 12 May 2021 02:37:16 +0000
Received: from pps.filterd (userp3020.oracle.com [127.0.0.1])
        by userp3020.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 14C2a2jr028938;
        Wed, 12 May 2021 02:37:15 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2177.outbound.protection.outlook.com [104.47.55.177])
        by userp3020.oracle.com with ESMTP id 38fh3xnkqu-2
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 12 May 2021 02:37:15 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=eIjo3M/hrMClrckxFFAz3SI6EFA9rUmZ4oslO7/cRh33/RPdxLsMBOMxWbjKsQkU+/sE+t5GPJYzmHZ5GGgRz0KE4mZmO4KshojoY+EzjfYxgmBSYa3RHCdXoAsGNuGYFxA1ahfJ1h1FwsLgu3yH1htosfjDY3A3loC7ERgqzKdRJPFDdxY9cM8hnkhu9cuC1M0l92zKp/qDpb2/UogmuAxI3zV79bYntnfdLr6TuuHuNyRfY8q8EQGWQMRIWjkYcAey03NUMzOUU1BjsoEdi5olJoWBxOjlaBFQ0hLnPmMdG8l0tfpeuybgDycR7Femd3/sWlOFBpH6BqYwXpJFmQ==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=lYaG9mCpHdaSAqGNzPGndXNZxWV9gxCPg2KmqGBrnMA=;
 b=erXF6Msq3PVpCjeUANM/RlVJYVEwW0berRoV6vpJRjXUPubBDasu0FT4fQRA9KZUkTE+4ho6ASinukyXaneq5ZJ9Fe0YjbGw3HrHoO3uHgF9PXslLXaKL4xotjEzcsiccJ+WERi44jOkUYzCpz7cR1aNkzTEADWOlSbRiC5ykNU1dkVPVDJtAB3i06giDsCKUBrOAnuJuwD+jHtwb/m8+ma3XmULd6eK2ydRAe5uOiKb2mmWxEdr30kAmJfzn234SYXF0hDW1HWnr1ERF8/ACvm68PCTTLeJFH8CdI5pZOsiMdqTZ2xqljtTW7cxC8rxmz2w7QRTIl3KA06lgRHnlQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=lYaG9mCpHdaSAqGNzPGndXNZxWV9gxCPg2KmqGBrnMA=;
 b=MQUm5ixs7LUWdIzCth7TEN7rw0MmnAwjZ1+kQIo+FNrmZ3cWDeI6tqi0RGNnR65CUGvu1t3DXqgkdoE5pvXS21y+3N+BKAVCui8vB8wjWQOIDKK+45ei8EEVU2My04AdUlc8P/SUKD+ZNkQYvzr1+YDIowsCtdbfhUr3stmwmag=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4425.namprd10.prod.outlook.com (2603:10b6:806:11b::9) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4129.26; Wed, 12 May
 2021 02:37:13 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::5911:9489:e05c:2d44]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::5911:9489:e05c:2d44%5]) with mapi id 15.20.4108.031; Wed, 12 May 2021
 02:37:13 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 3/3] KVM: x86: Add a new VM statistic to show number of VCPUs
 created in a given VM
Date: Tue, 11 May 2021 21:47:59 -0400
Message-Id: <20210512014759.55556-4-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210512014759.55556-1-krish.sadhukhan@oracle.com>
References: <20210512014759.55556-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.201.29]
X-ClientProxiedBy: BY5PR03CA0016.namprd03.prod.outlook.com
 (2603:10b6:a03:1e0::26) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.201.29) by
 BY5PR03CA0016.namprd03.prod.outlook.com (2603:10b6:a03:1e0::26) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4108.25 via Frontend
 Transport; Wed, 12 May 2021 02:37:12 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: bd171759-a2e8-4b40-11a6-08d914eed9aa
X-MS-TrafficTypeDiagnostic: SA2PR10MB4425:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB44258660D6BF5720BA54165C81529@SA2PR10MB4425.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:4303;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 IIFOBRoSufUZFu7sqsgXTmRUTbVPecYGkka4HOOxLONR2ttfkicWCDub+hfbg1KUxIMaEJ0d31mPFnxFMOe0+SN7Ocz9E5pY8JCsfvu8rfobRJfqG41dU/D5oEzTaGDL7+AFY0WwkfjKOHZX5PcBuSUKgXEm2u8c3Pshp0GMkQARk8ooEBN14zcJ51cSiVlN+WN5D2vDe/lCP8dgkOyU0vqQRS2utFiQLQRl9lsBZeZXs6mQ1xTmeCl1esdoxckMUfTHsAFDSwKSKgZJHB75LWEBjnCh0Fw6qCe4Z+lZdb7QmDbinvXJXsN5jhLLKvKmj1vl0ZBuoizVy+zlxP8mYnWQYJtI3u1FTC++Nw/hQOlVwzMublhPvzRtmSko0iN5f41I3tAz1izoe0M5UxOBaBjV8bDer9CpVrBdKBF3U3xuDh1d/QI9WJXVbIoMmgsmGseFWdVseiuf9rKh30YYQTQX0XScEuP8tLO9BMo2T0IPIrDOP1cRm2OdbpwdX5L/0zpBU0QHk1yyqtMbPffrYoElg4zOaQUTnm3y0tjdbfwPFVRPjeM17yjA7gM4bTfocODM3FQL5A4bdO1zI6UH6BzBkfSNKlJzBLD+cmx4moNfzXr76DF2roH37A+M/g10/PHIjsANj+9WCVZiiupNDg==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(346002)(136003)(39860400002)(366004)(396003)(376002)(66946007)(8936002)(66556008)(478600001)(6486002)(2906002)(1076003)(36756003)(5660300002)(83380400001)(186003)(16526019)(66476007)(316002)(86362001)(6666004)(52116002)(38100700002)(8676002)(26005)(4326008)(38350700002)(6916009)(2616005)(44832011)(7696005)(956004);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 JuiQRcgrk1r3okrDG/ORsEgsTKNl7F1PpAx9dRUrlmJxzr5ED934aLgccok5xTd3yFDS5Ck/qzinKd5lvcbi8uAzTDVf1HfSRyku0w9HfQHZmWLAYRxju4mZWKF7/9AQOhVGgb/HzGiZK17AtfEs9NkdmP2D/4q7XoUfdaUmGEIg/4SakQQYRe23Hk9Z2PWD70fmNVbjCjFM+/swhLF4EKS3I0V/pJYb7C9LtPGKkcL9pjGVZzQJ9ACwfNOSBFpik9/O308zLXlMwU6Q/luGIbFBghJxwaSzuhRuaKga4D7kYAEydMr492aqRcheZ153tai/xQsH7H+YBjqQ4wwtXzd8smjUZEoaA+C4UAIfCr/vM27EyAJj0J56KugVYke4ZJ4PAp2tie649xcktLKp5iTl5DJM+LUuEGZfiXvGdFkFx10QXlA/EQzGK+cKR2lJyij0EDeu2IcLeyXaE/ARz7lZ3Pw2R9toVjDZLjsOnguczgW/iah+54J3SOkc2fMh+s/nCpRlqT7xPpXVFsOcCvBxA1+0cNG//kYWXzhq2e7eEP4lzjTH0/k+s3yBt5wmk0+fcUnz5SpDf8NUxQ4rJTS59XFrbWL1LWrLRtB/YoKnIt6CE95tSHo8PuYIajizffwzygpl+6gZ+QCgESgjgWgFWsYt2dwme5LsRERKnZINpm2w8YMU3YiPN1wRxsNR5zbFdr1aKsaEvOv3d0fp7UMZ2QdTxpl2zMXMs623C33WIKw3mC/G6s/vwx1cNz/t1LlPQkeI5bAabBaq39cjTXcP1GkUxXiVPpeSeaDLEmYHMXnF2Q4FNQyPbOceOl1W3ofGqwBjwtDtNYvN9TUE1sFWDhsf5vZiXmQw8QD0lu591VuI0EVu221Mfnw7U2mjjRO9VHbwtfgjIyDFSsbECwIrcveOAtRfPh3mrcXkPRZk4Swcer+51jTvxZbCQlssXfSWv2IYSLfYzcBZiIeORruTkjOCip1/NLQB0aINH/svlKfkxXcUIFwEqh+eoXJaNvNyvdjf44psiQfBCtBHj05Tb7BIf1SgjghA3QHoRqpleDPr/tWtVr5mDzsFIwaRQ5uBlFcleuRRFkFdnoxgrogX9lRAgPy3LjHNMkMQDkNaMBTkGTayeSJYZqXZUCbwg0aDLjp44AYiVpG24MHe5CsXF6ybAQ4OptNZwZfznsKrssOc9KCLokK/S3K5USRM9IwOcWU3szm4G2++sJ9vwAsvyAOEDZqYzcsre1sNMOsQAf4cSIAsga1QqtCp9A6muk5ILZANKT+MEzs+07NvPCLphd1zj10Y4o/tei5qo3hjBNEvsBiwFAsmC+2pE6ba
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 bd171759-a2e8-4b40-11a6-08d914eed9aa
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 12 May 2021 02:37:13.8268
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 P1MwvIfu5pLWg9No+vRdJG1PD+n5mk3vjZzBIlyYn8kiKfr92KRCRqJ2byndF3uo5gyo4kWquXO3qQwXYE8wCiyVzhl9+sHd0qXGq1HRjfo=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4425
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9981
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 spamscore=0
 phishscore=0 suspectscore=0
 malwarescore=0 adultscore=0 bulkscore=0 mlxlogscore=999 mlxscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2104190000
 definitions=main-2105120018
X-Proofpoint-GUID: Z6fANM9MX-S2ULKsS3V3EoxrhxAeRmVZ
X-Proofpoint-ORIG-GUID: Z6fANM9MX-S2ULKsS3V3EoxrhxAeRmVZ
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9981
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0
 lowpriorityscore=0 malwarescore=0
 bulkscore=0 spamscore=0 clxscore=1015 priorityscore=1501 adultscore=0
 mlxlogscore=999 mlxscore=0 suspectscore=0 impostorscore=0 phishscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2104190000
 definitions=main-2105120018
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

'struct kvm' already has a member for counting the number of VCPUs created
for a given VM. Add this as a new VM statistic to KVM debugfs.

Signed-off-by: Krish Sadhukhan <Krish.Sadhukhan@oracle.com>
---
 arch/x86/include/asm/kvm_host.h | 1 +
 arch/x86/kvm/x86.c              | 1 +
 virt/kvm/kvm_main.c             | 2 ++
 3 files changed, 4 insertions(+)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index 884f6e5ba669..4a3e3c04ef38 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1138,6 +1138,7 @@ struct kvm_vm_stat {
 	ulong lpages;
 	ulong nx_lpage_splits;
 	ulong max_mmu_page_hash_collisions;
+	u64 created_vcpus;
 };
 
 struct kvm_vcpu_stat {
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 01805b68dc9b..0c2a57e1e096 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -258,6 +258,7 @@ struct kvm_stats_debugfs_item debugfs_entries[] = {
 	VM_STAT("largepages", lpages, .mode = 0444),
 	VM_STAT("nx_largepages_splitted", nx_lpage_splits, .mode = 0444),
 	VM_STAT("max_mmu_page_hash_collisions", max_mmu_page_hash_collisions),
+	VM_STAT("created_vcpus", created_vcpus),
 	{ NULL }
 };
 
diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 6b4feb92dc79..ac8f02d8a051 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -3318,6 +3318,7 @@ static int kvm_vm_ioctl_create_vcpu(struct kvm *kvm, u32 id)
 	}
 
 	kvm->created_vcpus++;
+	kvm->stat.created_vcpus++;
 	mutex_unlock(&kvm->lock);
 
 	r = kvm_arch_vcpu_precreate(kvm, id);
@@ -3394,6 +3395,7 @@ static int kvm_vm_ioctl_create_vcpu(struct kvm *kvm, u32 id)
 vcpu_decrement:
 	mutex_lock(&kvm->lock);
 	kvm->created_vcpus--;
+	kvm->stat.created_vcpus--;
 	mutex_unlock(&kvm->lock);
 	return r;
 }
